File: packages/meeting-detector/src/detector.ts
import { spawn, ChildProcess } from 'node:child_process';
import { EventEmitter } from 'node:events';
import { fileURLToPath } from 'node:url';
import { dirname, join } from 'node:path';
import { MeetingSignal, MeetingDetectorOptions, MeetingEventCallback, ErrorEventCallback } from './types.js';

interface SessionInfo {
  lastSeen: number;
  signal: MeetingSignal;
}

interface PendingConfidenceSignal {
  firstSeen: number;
  lastSeen: number;
  count: number;
  signal: MeetingSignal;
}

interface ServiceContext {
  frontApp?: string;
  windowTitle?: string;
}

export class MeetingDetector extends EventEmitter {
  private static readonly LOW_CONFIDENCE_WINDOW_MS = 45000;
  private static readonly LOW_CONFIDENCE_FALLBACK_MIN_SIGNALS = 4;
  private static readonly LOW_CONFIDENCE_FALLBACK_MIN_DURATION_MS = 30000;
  private static readonly PRECHECK_PRONE_SERVICES = new Set([
    'microsoft teams',
    'zoom',
    'webex',
    'slack'
  ]);

  private process?: ChildProcess;
  private options: Required<MeetingDetectorOptions>;
  private activeSessions: Map<string, SessionInfo> = new Map();
  private pendingConfidence: Map<string, PendingConfidenceSignal> = new Map();
  private serviceContext: Map<string, ServiceContext> = new Map();

  constructor(options: MeetingDetectorOptions = {}) {
    super();

    // Get the absolute path to the script relative to this package
    const defaultScriptPath = options.scriptPath || join(
      dirname(fileURLToPath(import.meta.url)),
      '../meeting-detect.sh'
    );

    this.options = {
      scriptPath: defaultScriptPath,
      debug: options.debug || false,
      sessionDeduplicationMs: options.sessionDeduplicationMs || 60000
    };
  }

  /**
   * Start monitoring for meeting signals
   * @param callback Optional callback function for meeting events
   */
  public start(callback?: MeetingEventCallback): void {
    if (this.process) {
      throw new Error('Detector is already running');
    }

    if (callback) {
      this.on('meeting', callback);
    }

    this.process = spawn('sh', [this.options.scriptPath], {
      stdio: ['pipe', 'pipe', 'pipe']
    });

    this.process.stdout?.on('data', (data: Buffer) => {
      const lines = data.toString().trim().split('\n');

      for (const line of lines) {
        if (line.trim()) {
          try {
            const parsedSignal = this.parseSignal(line);
            const signal = this.stabilizeSignalContext(parsedSignal);
            if (this.shouldIgnoreSignal(signal)) {
              if (this.options.debug) {
                console.log('[MeetingDetector] Ignoring signal:', signal);
              }
              continue;
            }

            const confidentSignal = this.resolveConfidence(signal);
            if (!confidentSignal) {
              if (this.options.debug) {
                console.log('[MeetingDetector] Holding low-confidence signal:', signal);
              }
              continue;
            }

            if (this.isDuplicateSession(confidentSignal)) {
              if (this.options.debug) {
                console.log('[MeetingDetector] Skipping duplicate session:', confidentSignal);
              }
              continue;
            }

            if (this.options.debug) {
              console.log('[MeetingDetector] Parsed signal:', confidentSignal);
            }
            this.emit('meeting', confidentSignal);
          } catch (error) {
            if (this.options.debug) {
              console.log('[MeetingDetector] Failed to parse line:', line);
            }
            this.emit('error', new Error(`Failed to parse signal: ${line}`));
          }
        }
      }
    });

    this.process.stderr?.on('data', (data: Buffer) => {
      if (this.options.debug) {
        console.log('[MeetingDetector] stderr:', data.toString());
      }
    });

    this.process.on('error', (error) => {
      this.emit('error', error);
    });

    this.process.on('exit', (code, signal) => {
      if (this.options.debug) {
        console.log(`[MeetingDetector] Process exited with code ${code}, signal ${signal}`);
      }
      this.process = undefined;
      this.emit('exit', { code, signal });
    });

    if (this.options.debug) {
      console.log('[MeetingDetector] Started monitoring');
    }
  }

  /**
   * Stop monitoring
   */
  public stop(): void {
    if (this.process) {
      this.process.kill('SIGTERM');
      this.process = undefined;

      // Clear active sessions when stopping
      this.activeSessions.clear();
      this.pendingConfidence.clear();
      this.serviceContext.clear();

      if (this.options.debug) {
        console.log('[MeetingDetector] Stopped monitoring');
      }
    }
  }

  /**
   * Check if the detector is currently running
   */
  public isRunning(): boolean {
    return !!this.process;
  }

  /**
   * Add a meeting event listener
   */
  public onMeeting(callback: MeetingEventCallback): void {
    this.on('meeting', callback);
  }

  /**
   * Add an error event listener
   */
  public onError(callback: ErrorEventCallback): void {
    this.on('error', callback);
  }

  /**
   * Comprehensive filtering to prevent false positives
   * Filters out:
   * - System processes (WebKit, SiriNCService, Chrome Helper, etc.)
   * - Development tools (Electron, Terminal, Xcode)
   * - Generic browser services (Chrome, Safari, Firefox)
   * - Browser camera initialization (no window title + 'requested' verdict)
   * - Google Meet signals without valid meeting URL patterns
   */
  private shouldIgnoreSignal(signal: MeetingSignal): boolean {
    // System processes that should never trigger meeting detection
    const systemProcessPatterns = [
      'sirinc',            // SiriNCService
      'afplay',            // macOS audio file player
      'systemsoundserver', // System sound effects
      'wavelink',          // Audio routing software
      'granola helper',    // Screen recording helper
      'webkit.gpu',        // WebKit GPU process
      'webkit.networking', // WebKit networking
      // NOTE: 'chrome helper' intentionally NOT listed — Chrome Helper is the process
      // used by Google Meet, Zoom web, and other browser-based meeting platforms.
      'electron helper',   // Electron helper processes (not meeting-specific)
      'caphost',           // Zoom internal media helper (emitted separately)
      'webview helper',    // Generic WKWebView helper
    ];

    // Services/apps that are too generic or non-meeting
    const genericServices = [
      'electron',
      'terminal',
      'granola',
      'finder',
      'xcode',
      'tips',
      'google chrome',    // Generic Chrome (resolved to 'Google Meet' when in a call)
      'safari',
      'firefox',
      'microsoft edge',
      'photo booth',
      'quicktime player',
      'quicktime playerx',
    ];

    const processName = signal.process?.toLowerCase() || '';
    const serviceName = signal.service?.toLowerCase() || '';

    // Filter by process name patterns (partial match)
    if (systemProcessPatterns.some(pattern => processName.includes(pattern))) {
      return true;
    }

    // Filter by exact generic service names
    if (genericServices.includes(serviceName)) {
      return true;
    }

    // Camera initialization filter: if verdict is 'requested' and window_title is empty
    // AND the camera hardware is not yet active, it's a pre-check, not an active call.
    if (signal.verdict === 'requested' && (!signal.window_title || signal.window_title.trim() === '')) {
      if (!signal.camera_active) {
        return true;
      }
    }

    // For Google Meet (Chrome-based): validate title only when it is available.
    // When Chrome is backgrounded the title is empty — still allow the signal through
    // because the meeting is genuinely active (camera_active guard above already handles
    // the pre-check case).
    if (serviceName === 'google meet') {
      const windowTitle = signal.window_title?.trim() || '';
      if (windowTitle !== '') {
        // Title present → require it to look like an actual meeting room
        const hasValidMeetTitle =
          windowTitle.includes('meet.google.com') ||
          windowTitle.includes('Meet - ') ||
          /[a-z]{3}-[a-z]{4}-[a-z]{3}/.test(windowTitle);
        if (!hasValidMeetTitle) {
          return true;
        }
      }
      // Empty title → pass through (Chrome is backgrounded but meeting is active)
    }

    return false;
  }

  private getServiceKey(signal: MeetingSignal): string {
    return (signal.service || signal.front_app || signal.process || '').toLowerCase();
  }

  private isFrontAppConsistentWithService(frontApp: string, service: string): boolean {
    const f = (frontApp || '').toLowerCase();
    const s = (service || '').toLowerCase();
    if (!f || !s) return false;

    if (s === 'microsoft teams') {
      return f.includes('teams') || f.includes('msteams');
    }
    if (s === 'google meet') {
      return f.includes('chrome') || f.includes('google meet');
    }
    if (s === 'zoom') {
      return f.includes('zoom');
    }
    if (s === 'webex') {
      return f.includes('webex');
    }
    if (s === 'slack') {
      return f.includes('slack');
    }
    return f.includes(s);
  }

  private stabilizeSignalContext(signal: MeetingSignal): MeetingSignal {
    const key = this.getServiceKey(signal);
    const existing = this.serviceContext.get(key) || {};
    const next: MeetingSignal = { ...signal };

    const rawFront = (signal.front_app || '').trim();
    const rawTitle = (signal.window_title || '').trim();
    const frontConsistent = this.isFrontAppConsistentWithService(rawFront, signal.service);

    if (rawFront && frontConsistent) {
      next.front_app = rawFront;
      existing.frontApp = rawFront;
    } else if (existing.frontApp) {
      next.front_app = existing.frontApp;
    } else {
      // Keep context aligned with detected service when OS foreground sampling is stale.
      next.front_app = signal.service;
      existing.frontApp = signal.service;
    }

    if (rawTitle && frontConsistent) {
      next.window_title = rawTitle;
      existing.windowTitle = rawTitle;
    } else if (existing.windowTitle && this.isFrontAppConsistentWithService(next.front_app, signal.service)) {
      next.window_title = existing.windowTitle;
    } else {
      next.window_title = '';
    }

    this.serviceContext.set(key, existing);
    return next;
  }

  private isLowConfidenceSignal(signal: MeetingSignal): boolean {
    const serviceKey = this.getServiceKey(signal);
    if (!MeetingDetector.PRECHECK_PRONE_SERVICES.has(serviceKey)) {
      return false;
    }

    const hasNoWindowTitle = !signal.window_title || signal.window_title.trim() === '';
    return signal.verdict === 'requested' && signal.preflight === true && hasNoWindowTitle;
  }

  private hasStrongMeetingEvidence(signal: MeetingSignal): boolean {
    if (signal.verdict === 'allowed' || signal.preflight === false) {
      return true;
    }
    return !!signal.window_title && signal.window_title.trim() !== '';
  }

  private cleanupExpiredPendingConfidence(now: number): void {
    for (const [key, pending] of this.pendingConfidence.entries()) {
      if (now - pending.lastSeen > MeetingDetector.LOW_CONFIDENCE_WINDOW_MS) {
        this.pendingConfidence.delete(key);
      }
    }
  }

  private resolveConfidence(signal: MeetingSignal): MeetingSignal | null {
    const now = Date.now();
    this.cleanupExpiredPendingConfidence(now);

    const key = this.getServiceKey(signal);
    const pending = this.pendingConfidence.get(key);
    const lowConfidence = this.isLowConfidenceSignal(signal);
    const strongEvidence = this.hasStrongMeetingEvidence(signal);

    if (strongEvidence) {
      this.pendingConfidence.delete(key);
      return signal;
    }

    if (!lowConfidence) {
      return signal;
    }

    if (!pending) {
      this.pendingConfidence.set(key, {
        firstSeen: now,
        lastSeen: now,
        count: 1,
        signal
      });
      return null;
    }

    pending.lastSeen = now;
    pending.count += 1;
    pending.signal = signal;

    const duration = now - pending.firstSeen;
    if (
      pending.count >= MeetingDetector.LOW_CONFIDENCE_FALLBACK_MIN_SIGNALS &&
      duration >= MeetingDetector.LOW_CONFIDENCE_FALLBACK_MIN_DURATION_MS
    ) {
      this.pendingConfidence.delete(key);
      if (this.options.debug) {
        console.log('[MeetingDetector] Promoting low-confidence signal after sustained activity:', signal);
      }
      return signal;
    }

    return null;
  }

  /**
   * Generate a unique session key based on the signal properties
   */
  private getSessionKey(signal: MeetingSignal): string {
    // Session key is service-centric to collapse helper-process bursts from the same app.
    return signal.service || signal.front_app || signal.process;
  }

  /**
   * Check if this signal is a duplicate of an existing session
   * Returns true if duplicate, false if new or expired session
   */
  private isDuplicateSession(signal: MeetingSignal): boolean {
    const sessionKey = this.getSessionKey(signal);
    const now = Date.now();
    const existingSession = this.activeSessions.get(sessionKey);

    // Clean up expired sessions periodically
    this.cleanupExpiredSessions();

    if (existingSession) {
      const timeSinceLastSeen = now - existingSession.lastSeen;

      if (timeSinceLastSeen < this.options.sessionDeduplicationMs) {
        // Update last seen time for existing session
        existingSession.lastSeen = now;
        return true; // This is a duplicate
      }
    }

    // New session or expired session - track it
    this.activeSessions.set(sessionKey, {
      lastSeen: now,
      signal
    });

    return false; // This is not a duplicate
  }

  /**
   * Clean up sessions that have expired
   */
  private cleanupExpiredSessions(): void {
    const now = Date.now();
    const expiredKeys: string[] = [];

    for (const [key, session] of this.activeSessions.entries()) {
      if (now - session.lastSeen > this.options.sessionDeduplicationMs) {
        expiredKeys.push(key);
      }
    }

    for (const key of expiredKeys) {
      this.activeSessions.delete(key);
      if (this.options.debug) {
        console.log(`[MeetingDetector] Cleaned up expired session: ${key}`);
      }
    }
  }

  private parseSignal(line: string): MeetingSignal {
    const signal = JSON.parse(line) as Record<string, any>;
    const chromeUrl = signal.chrome_url || '';

    // Use transformed app name as service if the original service is a system service like 'microphone' or 'camera'
    const originalService = signal.service || '';
    const transformedService = this.transformAppName(signal.front_app, signal.process, chromeUrl);
    const finalService = (originalService === 'microphone' || originalService === 'camera' || !originalService)
      ? transformedService
      : originalService;

    return {
      event: signal.event,
      timestamp: signal.timestamp,
      service: finalService,
      verdict: signal.verdict || '',
      preflight: signal.preflight === 'true' || signal.preflight === true,
      process: signal.process || '',
      pid: signal.pid || '',
      parent_pid: signal.parent_pid || '',
      process_path: signal.process_path || '',
      front_app: signal.front_app || '',
      window_title: signal.window_title || '',
      session_id: signal.session_id || '',
      camera_active: signal.camera_active === 'true' || signal.camera_active === true,
      chrome_url: chromeUrl
    };
  }

  private includesAny(value: string, patterns: string[]): boolean {
    return patterns.some((pattern) => value.includes(pattern));
  }

  private transformAppName(frontApp: string, process: string, chromeUrl = ''): string {
    const app = frontApp?.toLowerCase() || '';
    const proc = process?.toLowerCase() || '';
    const url = chromeUrl.toLowerCase();

    // For Chrome Helper processes, the active tab URL is the definitive source —
    // it does not depend on which app is currently frontmost.
    if (url && proc.includes('chrome')) {
      if (url.includes('meet.google.com')) return 'Google Meet';
      if (url.includes('zoom.us/wc/') || url.includes('zoom.us/j/')) return 'Zoom';
      if (url.includes('teams.microsoft.com') || url.includes('teams.live.com')) return 'Microsoft Teams';
      if (url.includes('web.webex.com') || url.includes('webex.com/meet')) return 'Webex';
      if (url.includes('app.slack.com') && url.includes('huddle')) return 'Slack';
      if (url.includes('meet.jit.si') || url.includes('jitsi')) return 'Jitsi Meet';
      if (url.includes('whereby.com')) return 'Whereby';
      if (url.includes('bluejeans.com')) return 'BlueJeans';
      if (url.includes('ringcentral.com')) return 'RingCentral';
      if (url.includes('chime.aws')) return 'Amazon Chime';
      if (url.includes('goto.com') || url.includes('gotomeeting.com')) return 'GoToMeeting';
    }

    // Prefer process identity next because front_app sampling can be stale.
    if (this.includesAny(proc, ['microsoft teams', 'msteams'])) return 'Microsoft Teams';
    if (this.includesAny(proc, ['zoom'])) return 'Zoom';
    if (this.includesAny(proc, ['webex', 'cisco webex'])) return 'Webex';
    if (this.includesAny(proc, ['slack'])) return 'Slack';
    if (this.includesAny(proc, ['google meet', 'meet.google.com'])) return 'Google Meet';
    if (this.includesAny(proc, ['skype'])) return 'Skype';
    if (this.includesAny(proc, ['discord'])) return 'Discord';
    if (this.includesAny(proc, ['facetime'])) return 'FaceTime';
    if (this.includesAny(proc, ['gotomeeting', 'goto meeting'])) return 'GoToMeeting';
    if (this.includesAny(proc, ['bluejeans', 'blue jeans'])) return 'BlueJeans';
    if (this.includesAny(proc, ['jitsi'])) return 'Jitsi Meet';
    if (this.includesAny(proc, ['whereby'])) return 'Whereby';
    if (this.includesAny(proc, ['8x8'])) return '8x8';
    if (this.includesAny(proc, ['ringcentral', 'ring central'])) return 'RingCentral';
    if (this.includesAny(proc, ['bigbluebutton', 'big blue button'])) return 'BigBlueButton';
    if (this.includesAny(proc, ['amazon chime', 'chime'])) return 'Amazon Chime';
    if (this.includesAny(proc, ['google hangouts', 'hangouts'])) return 'Google Hangouts';
    if (this.includesAny(proc, ['adobe connect'])) return 'Adobe Connect';
    if (this.includesAny(proc, ['teamviewer'])) return 'TeamViewer';
    if (this.includesAny(proc, ['anydesk'])) return 'AnyDesk';
    if (this.includesAny(proc, ['clickmeeting'])) return 'ClickMeeting';
    if (this.includesAny(proc, ['appear.in'])) return 'Appear.in';

    // Fallback to front_app when process identity is generic/indirect (e.g., Chrome Helper
    // without a chrome_url resolved). front_app is unreliable when Chrome is backgrounded
    // — only use it when we have no better signal.
    if (this.includesAny(app, ['microsoft teams', 'msteams'])) return 'Microsoft Teams';
    if (this.includesAny(app, ['zoom'])) return 'Zoom';
    if (this.includesAny(app, ['webex'])) return 'Webex';
    if (this.includesAny(app, ['slack'])) return 'Slack';
    if (this.includesAny(app, ['google meet'])) return 'Google Meet';
    if (this.includesAny(app, ['chrome'])) return 'Google Meet';
    if (this.includesAny(app, ['skype'])) return 'Skype';
    if (this.includesAny(app, ['discord'])) return 'Discord';
    if (this.includesAny(app, ['facetime'])) return 'FaceTime';
    if (this.includesAny(app, ['gotomeeting'])) return 'GoToMeeting';
    if (this.includesAny(app, ['bluejeans'])) return 'BlueJeans';
    if (this.includesAny(app, ['jitsi'])) return 'Jitsi Meet';
    if (this.includesAny(app, ['whereby'])) return 'Whereby';
    if (this.includesAny(app, ['8x8'])) return '8x8';
    if (this.includesAny(app, ['ringcentral'])) return 'RingCentral';
    if (this.includesAny(app, ['chime'])) return 'Amazon Chime';
    if (this.includesAny(app, ['hangouts'])) return 'Google Hangouts';
    if (this.includesAny(app, ['adobe connect'])) return 'Adobe Connect';
    if (this.includesAny(app, ['teamviewer'])) return 'TeamViewer';
    if (this.includesAny(app, ['anydesk'])) return 'AnyDesk';
    if (this.includesAny(app, ['clickmeeting'])) return 'ClickMeeting';
    if (this.includesAny(app, ['appear.in'])) return 'Appear.in';

    // Final fallback.
    return frontApp || 'Meeting App';
  }
}

// Convenience function for simple usage
export function detector(callback: MeetingEventCallback, options?: MeetingDetectorOptions): MeetingDetector {
  const detector = new MeetingDetector(options);
  detector.start(callback);
  return detector;
}

File: packages/meeting-detector/native/src/detector.rs
//! Platform-agnostic meeting detection state machine.
//!
//! This module handles the core detection logic:
//! - Session deduplication
//! - Confidence scoring
//! - Lifecycle event generation
//! - Meeting end timeout handling

use crate::error::{DetectorError, DetectorResult};
use crate::matchers::{MatchContext, MatcherRegistry};
use crate::types::{
    Confidence, DetectorOptions, LifecycleReason, MeetingLifecycleEvent,
    MeetingPlatform, MeetingSignal,
};
use std::collections::HashMap;
use std::sync::{Arc, Mutex};
use std::time::{Duration, Instant};

/// Session info for deduplication.
#[derive(Debug, Clone)]
struct SessionInfo {
    last_seen: Instant,
    signal: MeetingSignal,
}

/// Pending low-confidence signal.
#[derive(Debug, Clone)]
struct PendingConfidenceSignal {
    first_seen: Instant,
    last_seen: Instant,
    count: u32,
    signal: MeetingSignal,
}

/// Active meeting state.
#[derive(Debug, Clone)]
struct ActiveMeetingState {
    platform: MeetingPlatform,
    last_seen: Instant,
    confidence: Confidence,
    signal: MeetingSignal,
}

/// Configuration for the detector state machine.
#[derive(Debug, Clone)]
pub struct DetectorConfig {
    pub debug: bool,
    pub session_deduplication_ms: u64,
    pub meeting_end_timeout_ms: u64,
    pub emit_unknown: bool,
    pub include_sensitive_metadata: bool,
    pub include_raw_signal_in_lifecycle: bool,
    pub startup_probe: bool,
}

impl Default for DetectorConfig {
    fn default() -> Self {
        Self {
            debug: false,
            session_deduplication_ms: 60000,
            meeting_end_timeout_ms: 30000,
            emit_unknown: false,
            include_sensitive_metadata: false,
            include_raw_signal_in_lifecycle: false,
            startup_probe: true,
        }
    }
}

impl From<DetectorOptions> for DetectorConfig {
    fn from(opts: DetectorOptions) -> Self {
        Self {
            debug: opts.debug.unwrap_or(false),
            session_deduplication_ms: opts.session_deduplication_ms.unwrap_or(60000),
            meeting_end_timeout_ms: opts.meeting_end_timeout_ms.unwrap_or(30000),
            emit_unknown: opts.emit_unknown.unwrap_or(false),
            include_sensitive_metadata: opts.include_sensitive_metadata.unwrap_or(false),
            include_raw_signal_in_lifecycle: opts.include_raw_signal_in_lifecycle.unwrap_or(false),
            startup_probe: opts.startup_probe.unwrap_or(true),
        }
    }
}

/// Meeting detection state machine.
pub struct DetectorStateMachine {
    config: DetectorConfig,
    matcher_registry: MatcherRegistry,
    active_sessions: HashMap<String, SessionInfo>,
    pending_confidence: HashMap<String, PendingConfidenceSignal>,
    active_meeting: Option<ActiveMeetingState>,
    meeting_end_scheduled: Option<Instant>,
}

impl DetectorStateMachine {
    /// Low confidence window duration.
    const LOW_CONFIDENCE_WINDOW: Duration = Duration::from_millis(45000);
    /// Minimum signals for low confidence fallback.
    const LOW_CONFIDENCE_MIN_SIGNALS: u32 = 4;
    /// Minimum duration for low confidence fallback.
    const LOW_CONFIDENCE_MIN_DURATION: Duration = Duration::from_millis(30000);

    /// Services prone to preflight checks.
    const PRECHECK_PRONE_SERVICES: &'static [&'static str] = &[
        "microsoft teams",
        "zoom",
        "cisco webex",
        "slack",
    ];

    /// Create a new state machine.
    pub fn new(config: DetectorConfig) -> Self {
        Self {
            config,
            matcher_registry: MatcherRegistry::new(),
            active_sessions: HashMap::new(),
            pending_confidence: HashMap::new(),
            active_meeting: None,
            meeting_end_scheduled: None,
        }
    }

    /// Process an incoming signal.
    ///
    /// Returns lifecycle events generated from this signal.
    pub fn process_signal(&mut self, signal: MeetingSignal) -> Vec<MeetingLifecycleEvent> {
        let mut events = Vec::new();

        // Check if we should ignore this signal
        if self.should_ignore_signal(&signal) {
            if self.config.debug {
                eprintln!("[Detector] Ignoring signal: {:?}", signal.service);
            }
            return events;
        }

        // Resolve confidence
        let (confident_signal, confidence) = match self.resolve_confidence(signal) {
            Some(result) => result,
            None => return events,
        };

        // Check for duplicates
        if self.is_duplicate_session(&confident_signal) {
            if self.config.debug {
                eprintln!("[Detector] Skipping duplicate session");
            }
            return events;
        }

        // Match the platform
        let ctx = self.signal_to_match_context(&confident_signal);
        let platform = self.matcher_registry.match_platform(&ctx);

        // Skip unknown platforms unless configured to emit
        if platform == MeetingPlatform::Unknown && !self.config.emit_unknown {
            return events;
        }

        // Update lifecycle
        if let Some(event) = self.update_meeting_lifecycle(platform, confidence, &confident_signal) {
            events.push(event);
        }

        // Record this session
        self.record_session(&confident_signal);

        events
    }

    /// Check if the meeting end timeout has elapsed.
    pub fn check_meeting_end(&mut self) -> Option<MeetingLifecycleEvent> {
        if let Some(scheduled) = self.meeting_end_scheduled {
            if Instant::now() >= scheduled {
                self.meeting_end_scheduled = None;
                if let Some(meeting) = self.active_meeting.take() {
                    return Some(MeetingLifecycleEvent::meeting_ended(
                        meeting.platform,
                        meeting.confidence,
                        LifecycleReason::Timeout,
                    ));
                }
            }
        }
        None
    }

    /// Handle detector stop - emit meeting_ended if active.
    pub fn on_stop(&mut self) -> Option<MeetingLifecycleEvent> {
        self.meeting_end_scheduled = None;
        if let Some(meeting) = self.active_meeting.take() {
            return Some(MeetingLifecycleEvent::meeting_ended(
                meeting.platform,
                meeting.confidence,
                LifecycleReason::Stop,
            ));
        }
        None
    }

    /// Check if we should ignore this signal.
    fn should_ignore_signal(&self, signal: &MeetingSignal) -> bool {
        let process = signal.process.to_lowercase();
        let service = signal.service.to_lowercase();

        // System processes to ignore
        const SYSTEM_PROCESSES: &[&str] = &[
            "sirinc", "afplay", "systemsoundserver", "wavelink",
            "granola helper", "webkit.gpu", "webkit.networking",
            "electron helper", "caphost", "webview helper",
        ];

        // Generic services to ignore
        const GENERIC_SERVICES: &[&str] = &[
            "electron", "terminal", "granola", "finder", "xcode",
            "tips", "google chrome", "safari", "firefox",
            "microsoft edge", "photo booth", "quicktime player",
            "quicktime playerx",
        ];

        // Check system processes
        if SYSTEM_PROCESSES.iter().any(|p| process.contains(p)) {
            return true;
        }

        // Check generic services
        if GENERIC_SERVICES.contains(&service.as_str()) {
            return true;
        }

        // Skip unknown unless configured
        if service == "unknown" && !self.config.emit_unknown {
            return true;
        }

        // Camera initialization filter
        if signal.verdict == "requested" 
            && signal.window_title.trim().is_empty()
            && !signal.camera_active
        {
            return true;
        }

        false
    }

    /// Resolve signal confidence.
    fn resolve_confidence(&mut self, signal: MeetingSignal) -> Option<(MeetingSignal, Confidence)> {
        let service = signal.service.to_lowercase();
        let now = Instant::now();

        // High confidence signals pass through immediately
        if signal.camera_active || signal.verdict == "allowed" {
            // Clear any pending low-confidence for this service
            self.pending_confidence.remove(&service);
            return Some((signal, Confidence::High));
        }

        // Check if this service is prone to preflight checks
        let is_precheck_prone = Self::PRECHECK_PRONE_SERVICES
            .iter()
            .any(|s| service.contains(s));

        if !is_precheck_prone {
            return Some((signal, Confidence::Medium));
        }

        // Track low-confidence signals
        let pending = self.pending_confidence
            .entry(service.clone())
            .or_insert_with(|| PendingConfidenceSignal {
                first_seen: now,
                last_seen: now,
                count: 0,
                signal: signal.clone(),
            });

        pending.last_seen = now;
        pending.count += 1;
        pending.signal = signal;

        // Check if we've exceeded the confidence window
        let duration = now.duration_since(pending.first_seen);
        if duration > Self::LOW_CONFIDENCE_WINDOW {
            self.pending_confidence.remove(&service);
            return None;
        }

        // Check if we have enough signals
        if pending.count >= Self::LOW_CONFIDENCE_MIN_SIGNALS
            && duration >= Self::LOW_CONFIDENCE_MIN_DURATION
        {
            let result = pending.signal.clone();
            self.pending_confidence.remove(&service);
            return Some((result, Confidence::Low));
        }

        // Still collecting signals
        None
    }

    /// Check if this is a duplicate session.
    fn is_duplicate_session(&self, signal: &MeetingSignal) -> bool {
        if let Some(session) = self.active_sessions.get(&signal.session_id) {
            let elapsed = Instant::now().duration_since(session.last_seen);
            return elapsed.as_millis() < self.config.session_deduplication_ms as u128;
        }
        false
    }

    /// Record a session for deduplication.
    fn record_session(&mut self, signal: &MeetingSignal) {
        self.active_sessions.insert(
            signal.session_id.clone(),
            SessionInfo {
                last_seen: Instant::now(),
                signal: signal.clone(),
            },
        );
    }

    /// Convert signal to match context.
    fn signal_to_match_context(&self, signal: &MeetingSignal) -> MatchContext {
        MatchContext::new(&signal.process, &signal.window_title)
            .with_camera_active(signal.camera_active)
    }

    /// Update meeting lifecycle state.
    fn update_meeting_lifecycle(
        &mut self,
        platform: MeetingPlatform,
        confidence: Confidence,
        signal: &MeetingSignal,
    ) -> Option<MeetingLifecycleEvent> {
        let now = Instant::now();

        // Cancel any scheduled meeting end
        self.meeting_end_scheduled = None;

        match &self.active_meeting {
            None => {
                // New meeting started
                self.active_meeting = Some(ActiveMeetingState {
                    platform,
                    last_seen: now,
                    confidence,
                    signal: signal.clone(),
                });
                Some(MeetingLifecycleEvent::meeting_started(platform, confidence))
            }
            Some(current) if current.platform != platform => {
                // Meeting platform changed
                let previous = current.platform;
                self.active_meeting = Some(ActiveMeetingState {
                    platform,
                    last_seen: now,
                    confidence,
                    signal: signal.clone(),
                });
                Some(MeetingLifecycleEvent::meeting_changed(platform, previous, confidence))
            }
            Some(_) => {
                // Same meeting, update last_seen
                if let Some(meeting) = &mut self.active_meeting {
                    meeting.last_seen = now;
                    meeting.signal = signal.clone();
                }
                // Schedule meeting end timeout
                self.meeting_end_scheduled = Some(
                    now + Duration::from_millis(self.config.meeting_end_timeout_ms)
                );
                None
            }
        }
    }

    /// Clean up old sessions.
    pub fn cleanup_sessions(&mut self) {
        let now = Instant::now();
        let timeout = Duration::from_millis(self.config.session_deduplication_ms);
        
        self.active_sessions.retain(|_, session| {
            now.duration_since(session.last_seen) < timeout
        });
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    fn test_signal(service: &str, camera_active: bool) -> MeetingSignal {
        MeetingSignal {
            event: "meeting_signal".to_string(),
            timestamp: chrono::Utc::now().to_rfc3339(),
            service: service.to_string(),
            verdict: if camera_active { "allowed".to_string() } else { "requested".to_string() },
            preflight: false,
            process: service.to_string(),
            pid: "123".to_string(),
            parent_pid: "1".to_string(),
            process_path: format!("/Applications/{}.app", service),
            front_app: service.to_string(),
            window_title: format!("{} Meeting", service),
            session_id: format!("session-{}", rand::random::<u32>()),
            camera_active,
            chrome_url: None,
        }
    }

    #[test]
    fn test_meeting_lifecycle() {
        let config = DetectorConfig::default();
        let mut machine = DetectorStateMachine::new(config);

        // Start a Zoom meeting
        let signal = test_signal("Zoom", true);
        let events = machine.process_signal(signal);

        assert_eq!(events.len(), 1);
        assert_eq!(events[0].event, "meeting_started");
        assert_eq!(events[0].platform, "Zoom");
    }

    #[test]
    fn test_ignore_system_processes() {
        let config = DetectorConfig::default();
        let mut machine = DetectorStateMachine::new(config);

        // System process should be ignored
        let mut signal = test_signal("SiriNCService", true);
        signal.process = "sirinc".to_string();
        let events = machine.process_signal(signal);

        assert!(events.is_empty());
    }

    #[test]
    fn test_meeting_change() {
        let config = DetectorConfig::default();
        let mut machine = DetectorStateMachine::new(config);

        // Start Zoom meeting
        let zoom_signal = test_signal("Zoom", true);
        let _ = machine.process_signal(zoom_signal);

        // Switch to Teams
        let mut teams_signal = test_signal("Microsoft Teams", true);
        teams_signal.session_id = "new-session".to_string();
        let events = machine.process_signal(teams_signal);

        assert_eq!(events.len(), 1);
        assert_eq!(events[0].event, "meeting_changed");
        assert_eq!(events[0].platform, "Microsoft Teams");
        assert_eq!(events[0].previous_platform, Some("Zoom".to_string()));
    }

    #[test]
    fn test_stop_emits_ended() {
        let config = DetectorConfig::default();
        let mut machine = DetectorStateMachine::new(config);

        // Start a meeting
        let signal = test_signal("Zoom", true);
        machine.process_signal(signal);

        // Stop detector
        let event = machine.on_stop();

        assert!(event.is_some());
        let event = event.unwrap();
        assert_eq!(event.event, "meeting_ended");
        assert_eq!(event.reason, "stop");
    }
}
